src/grammar.rs

```diff
@@ -48,19 +52,111 @@ impl Grammar {
         }
     }
 
-    /// Get iterator of the `Grammar`'s `Productions`s
+    /// Get iterator of the `Grammar`'s `Production`s
     pub fn productions_iter(&self) -> Iter {
         Iter {
             iterator: self.productions.iter(),
         }
     }
 
-    /// Get mutable iterator of the `Grammar`'s `Productions`s
+    /// Get mutable iterator of the `Grammar`'s `Production`s
     pub fn productions_iter_mut(&mut self) -> IterMut {
         IterMut {
             iterator: self.productions.iter_mut(),
         }
     }
+
+    fn eval_terminal(&self, term: &Term) -> Result<String, Error> {
+        match *term {
+            Term::Nonterminal(ref nt) => self.traverse(&nt),
+            Term::Terminal(ref t) => Ok(t.clone()),
+        }
+    }
+
+    fn traverse(&self, ident: &String) -> Result<String, Error> {
+        let stack_red_zone: usize = 32 * 1024; // 32KB
+        if stacker::remaining_stack() <= stack_red_zone {
+            return Err(Error::GenerateError(
+                String::from("Infinite loop detected!"),
+            ));
+        }
+
+        let nonterm = Term::Nonterminal(ident.clone());
+        let mut result = String::new();
```

Comment:
Definitely a small point, but what do you think about moving `result` down to where it is used? As I was reading the code I had assigned some mental effort to tracking its mutations and was surprised to find none until the very end.